Add Counter and Histogram Aggregators

ankit-bhargava opened the pull request:
This PR builds out SDK support for the [Counter and Histogram Aggregators](https://github.com/jmacd/opentelemetry-specification/blob/jmacd/draft_metric_sdk_spec/specification/metrics/sdk.md#aggregator-implementations) which combine updates to metric instruments in meaningful values.  Each aggregator stores exportable values in a vector format for ease of use further down the metrics data pipeline.   All aggregators are derived from a base Aggregator abstract class which holds functions and data relevant to all aggregators.  Additional data structures, such as a vector for histogram boundaries, are added on a case by case basis.  Aggregators are dependent on the Metric Instruments API  (pending approval in [PR # 161](https://github.com/open-telemetry/opentelemetry-cpp/pull/161)) and **will not compile without those files.**

Also included are tests for each aggregator which validate the correctness of aggregation in both single-threaded and concurrent situations.

Note: We templated aggregators to support a wider array of scalar types.  This templating prevents us from using out of line declarations as we normally would.  As a result, we left all implementation code in the header file.  If there are any suggestions to remedy this we will gladly make the change.  

This PR is part of a series of PRs to add a full implementation of metrics API and SDK support.

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:
```diff
@@ -0,0 +1,103 @@
+
+#include <variant>
+#include <vector>
+#include <mutex>
```
nit - is there a reason these are not alphabetically sorted? If not, I would suggest to sort them.

ankit-bhargava replied:
the format tools seem to reorder these

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:
```diff
@@ -0,0 +1,103 @@
+    Aggregator() = default;
+    
+    /**
+     * Recieves a captured value from the instrument and applies it to the current aggregator value.
```
```suggestion
     * Receives a captured value from the instrument and applies it to the current aggregator value.
```

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/counter_aggregator.h:
```diff
@@ -0,0 +1,104 @@
+    void update(T val) override
+    {
+        this->mu_.lock();
+        this->values_[0] += val;  // atomic operation
```
For int and long, is it possible to spin + CAS (compare and swap) to achieve better perf?
Not blocking this PR though.

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/counter_aggregator.h:
```diff
@@ -0,0 +1,104 @@
+        if (this->kind_ == other.kind_)
+        {
+            this->mu_.lock();
+            this->values_[0] += other.values_[0];  // atomic operation afaik
```
curious, what does `// atomic operation afaik` mean?

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,160 @@
+     */
+    void update(T val) override
+    {
+        
```
nit: remove empty line?

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,160 @@
+    HistogramAggregator(metrics_api::BoundInstrumentKind kind, std::vector<double> boundaries)
+    {
+        this->kind_ = kind;
+        boundaries_   = boundaries;
```
Do we want to validate if the boundaries is monotonic?

ankit-bhargava replied:
Added a check which throws an invalid_argument exception if provided boundaries are not sorted

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,160 @@
+            }
+        }
+        
+        // Alternate implementation with binary search
```
It'll be helpful to put a comment on the design decision - e.g. why would we choose linear search vs. binary search.

ankit-bhargava replied:
Included some background on the topic above the update function -- let me know if this is not enough and I'll gladly add a more thorough explanation.

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,160 @@
+        
+        for (int i = 0; i < bucketCounts_.size(); i++)
+        {
+            bucketCounts_[i] += other.bucketCounts_[i];
```
Who should be responsible of making sure the `other.bucketCounts` won't overflow? Do we in general want to check the buckets are the same before proceeding?

ankit-bhargava replied:
This is a good suggestion.  When you say " `other.bucketCounts` won't overflow," are you referring to a scenario in which adding the values together would exceed the capacity of the data type?

reyang replied:
For example, if `other.bucketCounts_.size() <= i`.

reyang replied:
There are two separate things:
1. index overflow while accessing the bucketCounts_.
2. value overflow while performing `+=`.

1 is more scary.

ankit-bhargava replied:
I preempted the possibility of an index out of bounds issue by ensuring that the boundary vectors of both aggregators are identical.  

For the value overflow, I can think of some hacky ways to do it, for example if abs(a)+abs(b) < 0 to indicate a wraparound.  I could also subtract from a variable set to int_max.  I feel like there has to be a native c++ tool/function which does this but I can't seem to find it...

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,160 @@
+    
+    /**
+     * Merges the values of two aggregators in a semantically accurate manner.
+     * A histogram aggregator can only be merged with another histogram aggregatos with the same boudnaries.
```
Theoretically we could merge a more accurate histogram into a less accurate one (e.g. merge [0, 1, 10, 100, 1000) into [0, 10, 1000). Probably not something we should cover here.

ankit-bhargava commented:
Addressed comments and rebased.  This has a dependency on PR #161.

ankit-bhargava commented:
Failing Bazel noexcept test although I was under the impression we were allowed to throw exceptions in SDK components?  Unsure why the exporter proto CMake test is failing.

ankit-bhargava commented:
Passing all tests except for "exporter proto" which is an external issue.

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:
```diff
@@ -0,0 +1,164 @@
+#include "opentelemetry/core/timestamp.h"
+#include "opentelemetry/metrics/instrument.h"
+#include "opentelemetry/version.h"
+#include <vector>
```
Move this line above (right below <mutex>).

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,206 @@
+     */
+    HistogramAggregator(metrics_api::InstrumentKind kind, std::vector<double> boundaries)
+    {
+        if (!std::is_sorted(boundaries.begin(),boundaries.end())){
```
```suggestion
        if (!std::is_sorted(boundaries.begin(), boundaries.end())){
```

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:
```diff
@@ -0,0 +1,206 @@
+     *
+     * Depending on the use case, a linear search or binary search based implementation may be preferred.
+     * In uniformly distributed datasets, linear search outperforms binary search until 512 buckets.  However,
+     * if the distribution is strongly skewed right ( for example server latency where most values may be <10ms
```
```suggestion
     * if the distribution is strongly skewed right (for example server latency where most values may be <10ms
```

reyang commented:
@ankit-bhargava would you fix the CI/Format issue?

ankit-bhargava commented:
@reyang 
Fixed all the formatting issues and rebased.  Seeing two runtime_context files which should not be part of this PR and I'm not sure why they've been added and if they're necessary? Feel free to delete these and merge (or I can delete them).
